Support for coverage of Privilged Architecture

MuhammadHammad001 opened the pull request:
# mnemonic flag support:
The `mnemonic` can be used to write a coverpoint that should target a instruction of interest only. For instance, **lr** or **sc**. Coverpoint can be written in the following pattern:
```objective-c
csr_comb:
    mnemonic == 'lr' : 0
```

Commit:
Support for read_csr added, mstatush error resolved, immediate and rs2 value error for csrs resolved

Commit:
Issues resolved for hitcovpt and any_hitcovpt when read_csr accessed

Commit:
update to check only the register of interest for the old_fn_csr_comb

MuhammadHammad001 commented:
# Update to check the registers only of interest for read_csr() function:
First, if we used any coverpoint in the following pattern to compare the old_value of the register with the current value, it used to compare the register values with every instruction which was not required and we had no way to detect that whether the coverpoint of interest hit because of the required instruction or not. Now, if the user uses the coverpoints of the following pattern, then the isac will only check the instructions of interest for the coverpoint.
```objective-c
    (pmpcfg0    & 0x80 == 0x80) and (old("pmpcfg0") & 0xFF) ^ (pmpcfg0 & 0xFF) == 0x00 and old("pmpcfg0") != 0: 0
```
### Explanation of the coverpoint pattern
The above coverpoint will check whether the value of the register `pmpcfg0` is changed or not while trying to change in the assembly. So, consider the following:
```objective-c
mem[X,0x800003B0] -> 0x9073
mem[X,0x800003B2] -> 0x3A03
[176] [M]: 0x800003B0 (0x3A039073) csrrw zero, pmpcfg0, t2
CSR pmpcfg0 -> 0x00000000
CSR pmpcfg0 <- 0x9F9F9F9F (input: 0xFFFFFFFF)


mem[X,0x800003B4] -> 0xB073
mem[X,0x800003B6] -> 0x3A03
[177] [M]: 0x800003B4 (0x3A03B073) csrrc zero, pmpcfg0, t2
CSR pmpcfg0 -> 0x9F9F9F9F
CSR pmpcfg0 <- 0x9F9F9F9F (input: 0x00000000)
```
As, we try to change the value of the pmpcfg0 but it is not changed because of the lock bit, therefore, the value in the **old_csr_regfile** will not change which we can check using the coverpoint:
```objective-c
old("pmpcfg0") & 0xFF) ^ (pmpcfg0 & 0xFF) == 0x00 
```

Commit:
label_to_addr and mode check support added

MuhammadHammad001 commented:
# Mode Check Support:
In this commit, I have added the support for checking the Mode in which the hart is currently executed. Consider the following part of the log:
```objective-c
mem[X,0x80000A84] -> 0x2403
mem[X,0x80000A86] -> 0x1E41
[205] [M]: 0x80000A84 (0x1E412403) lw fp, 484(sp)
mem[R,0x800031E4] -> 0x5BFDDB7D
x8 <- 0x5BFDDB7D
```
Previously, there was no mechanism to determine the execution mode of the instruction.
### Solution
Now, users can easily check the mode in which the hart executed the instruction using the following coverpoint:
```objective-c
    "mode == 'M' and ((pmpaddr2) >= 0x00000000) and ((pmpaddr2) <= 0xFFFFFFFF)": 0
```
This coverpoint ensures that the current instruction of interest is executed in the required mode.

# Check address of Label defined in the test:
Earlier, there was no support for checking the address of labels defined in the test. Now, users can employ the following coverpoint pattern to find the address of any labeled point in the test:
```objective-c
    "get_addr('rvtest_data') == 0x80000388": 0
``` 
### Usage
The `get_addr()` function can be valuable for users writing coverpoints, ensuring that any exceptions occurring during the test are specifically within the designated area of interest.

allenjbaum commented:
Without going into implementation details, this looks really good, and are necessary for many tests. 
Have tests been written, along with a cgf file, so we can be sure they actually work properly?

Commit:
physical address, virtual address, page table walk variables added, issue 57 implemented

MuhammadHammad001 commented:
# Virtual Memory Implementation
In this commit, I have implemented the requested variables for the Physical address, virtual address, page table walk addresses for the instructions and data accesses in the [issue # 57](https://github.com/riscv-software-src/riscv-isac/issues/57) by [Mr. Pawan](pawan.kumar@incoresemi.com) in the [RISC-V ISAC](https://github.com/riscv-software-src/riscv-isac).

## Implementation
Consider the following log which will be used as a reference to explain the implementation:
```objective-c
mem[X,0x800007D4] -> 0x0073
mem[X,0x800007D6] -> 0x3020
[444] [M]: 0x800007D4 (0x30200073) mret
CSR mstatus <- 0x00000080
ret-ing from M to S


mem[R,0x82002900] -> 0x20800401
mem[R,0x82001000] -> 0x200000CF
mem[X,0x800007D8] -> 0xE737
mem[X,0x800007DA] -> 0x0000
[445] [S]: 0x900007D8 (0x0000E737) lui a4, 14
x14 <- 0x0000E000


mem[X,0x800007DC] -> 0x0713
mem[X,0x800007DE] -> 0xEAD7
[446] [S]: 0x900007DC (0xEAD70713) addi a4, a4, 3757
x14 <- 0x0000DEAD


mem[X,0x800007E0] -> 0xA423
mem[X,0x800007E2] -> 0x00ED
[447] [S]: 0x900007E0 (0x00EDA423) sw a4, 8(s11)
mem[R,0x82002930] -> 0x20800401
mem[R,0x8200100C] -> 0x20800CCF
mem[0x8200311C] <- 0x0000DEAD

```
- As you may see in the log that the mode is switched to the **S-mode** using _mret_ instruction and address translation is enabled using the **satp** register. For this test, the **SV-32** is implemented and _level 0_ is used.

Previously, there was no way to access any of the page table walks, virtual address and physical address in case of virtual memory implementation.

## Solution and Explanation

Now, consider the following coverpoint for understanding:
```objective-c
check_virtual_memory:
  config:
    - check ISA:=regex(.*32.*); check ISA:=regex(.*I.*Zicsr.*);  def rvtest_mtrap_routine=True; option VM = SV32;'
  mnemonics:
    sw: 0
  val_comb:
    'len_dptw == 2 and dptw1a == 0x82002930 and dptw0a == 0x8200100C and ieva == 0x900007E0 and iepa == 0x800007E0 and len_iptw == 2 and iptw1a == 0x82002900 and iptw0a == 0x82001000': 0

```
whose purpose is to target the following instruction in the above log file:
```objective-c
mem[X,0x800007E0] -> 0xA423
mem[X,0x800007E2] -> 0x00ED
[447] [S]: 0x900007E0 (0x00EDA423) sw a4, 8(s11)
mem[R,0x82002930] -> 0x20800401
mem[R,0x8200100C] -> 0x20800CCF
mem[0x8200311C] <- 0x0000DEAD
```
First of all, if you want to enable the following variables you need to add a flag **option VM = SV(addressing scheme)** where addressing scheme in this case was SV-32. You need to add this flag under the **config ** label.
```objective-c
  config:
    - check ISA:=regex(.*32.*); check ISA:=regex(.*I.*Zicsr.*);  def rvtest_mtrap_routine=True; option VM = SV32;'
```
As, you may see in the coverpoint, there are multiple variables being used, I will write the purpose of every variable one by one:

1. **ieva**
Virtual address of the instruction access
2. **iepa**
Physical address of the instruction access
3. **depa**
Physical address of the data access
4. **ieva_align**
Alignment check for the ieva
5. **iepa_align**
Alignment check for the iepa
6. **depa_align**
Alignment check for the depa

#### Instruction access page table walks
Now, for the page table walks for the address access of the instructions:
**iptw0a,iptw1a,iptw2a,iptw3a,iptw4a**

These variables are used for the page table walks for the instruction. To check how many were done in your case, you may use the following variable:
**len_iptw**

#### Data access page table walks
Now, for the page table walks for the address access of the data:
**dptw0a,dptw1a,dptw2a,dptw3a,dptw4a**

State for data access variables is not stored as these walks will be tracked on every data access. Their purpose is same as the instruction ones expect they are used for the data access page table walks.

### Important
State of the variables for the page table walks of instruction access has also been stored which you may verify from the above **sw a4, 8(s11)** coverpoint. When the first page table walk is done, the state of these instruction page table walk address variables is not changed untill the mode is switched back to **M - mode** or the page table is accessed again making it easy for the user to track.

Commit:
Support for values from memory using mem_val

MuhammadHammad001 commented:
## Value at Memory Location
In this commit, I have added the support for reading the value stored at a specific address. Consider the following coverpoint:
```objective-c
    'mem_val(get_addr("rvtest_data"),4) == 0xBEEDCAFE': 0
```
As, you may see that **get_addr()** that is added in previous [commit](https://github.com/riscv-software-src/riscv-isac/pull/80#issuecomment-1818339542) was used to extract the address using the label is used in this coverpoint to get the address of **rvtest_data** and then the _4_ written after the comma represents the size of memory we want to extract from the memory location. In this case, I wanted to get 4 Bytes, therefore, 4 is written. This is a function defined in the csr_comb.

User may also write a specific address of his/her interest to check the value stored at that address.
```objective-c
'(mem_val(0x80000388, 2) == 0xBEEDCAFE)': 0
```

Commit:
PTE Permission access function added

MuhammadHammad001 commented:
# PTE Permission Access Function
In this commit, I have added a function named as `get_pte_per()` which will make sure that permissions are set for the required physical address and also make sure that the PPN of the address at which the pte is stored is same as the PPN of the page table so we are sure that the pte is stored at the required location.

The python function is given as:
```python
            def get_pte_per(pa, pte_addr, pgtb_addr, pte_size):
                if (pgtb_addr >> 12) == (pte_addr >> 12):
                    if ((pa >> 12) == (get_mem_val(pte_addr, pte_size) >> 10)):
                        return get_mem_val(pte_addr, pte_size) & 0x3FF
                    else:
                        return None
                else:
                    return None
```

The coverpoint for this functionality is given as:

```objective-c
check_for_the_isac:
  config:
    - check ISA:=regex(.*32.*); check ISA:=regex(.*I.*Zicsr.*);  def rvtest_mtrap_routine=True; option VM = SV32;
  mnemonics:
    sw: 0
    csrrw: 0
    csrrs: 0
    csrrc: 0
    lw: 0
  val_comb:
    'get_pte_per(get_addr("rvtest_data"),rs1_val + imm_val,get_addr("rvtest_slvl1_pg_tbl"),2) == 0xCF': 0
```

So, using this coverpoint we are sure that the pte is set for the **rvtest_data**

allenjbaum commented:
Also note that the level variable start is dependent on the SATP mode 
(and, eventually the VSATP as well; when Sail implements H-extension, we will need to be able to distinguish them)
but can end anywhere between start and 1.  So that means code must now SATP.mode for the first PTE access (only)
in order to correctly label it, and each following one is decremented by one, 
but any writes use the value of the preceding access and don't decrement level.

Also note that tests for  VM should make sure that TLBs are turned off in whichever model is the reference model (for these tests only)

Commit:
Issue with the option resolved

Commit:
get_pte_per updated to take only three arguments. Useless pte_size removed.

Commit:
Virtual memory implementation changed. Now, depends on satp register, content for iptw and dptw added, pte_prop_name check added

MuhammadHammad001 commented:
# VIrtual Memory Implementation:
Now, the coverpoints don't need a special lablel for the virtual memory else the track of all the virtual memory registers is kept by the isac itself using the **satp** register. 

1. Now, the contents of the page table walk access have also been added which user may access using the label `iptw0cont` or `dptw0cont` e.t.c

```objective-c
    get_mem_val(2181042244,4) == 545259727: 0 #dptw0
    get_mem_val(2181048576,4) == 545260545: 0 #dptw1
```
# PTE Permissions function updated to GET_PTE

The function to get the pte permissions has been updated to `get_pte` which you may use like the following coverpoint:

```objective-c
  val_comb:
    'get_pte(get_addr("rvtest_data"),rs1_val + imm_val,get_addr("rvtest_slvl1_pg_tbl")) == 0x800000CF': 0
```

# Function to check the specific permission added

Now, the function to check the specific permission of the pte using the `prop_name` has also been added which you may use like the following:

```objective-c
    get_pte_prop('U',get_addr("begin_signature"), rs1_val + imm_val, get_addr("rvtest_slvl1_pg_tbl")) == 1: 0

```

Commit:
macro support added

MuhammadHammad001 commented:
# Macros file Support Added
In this commit, I have added support for using a macro in the coverpoints.
Please consider the following macros/header file for the coverpoint:
```objective-c
common:
 shift_bit: 0x001
self:
 final_value: 0x012
```
The header file is written in the `yaml` format. Anything under the common label is available for all the cgf files that may include the header file. 
User needs to add the following flags while running the `risc-v isac` command:
```objective-c
     -h -> header file flag
    -cm -> cgf macro flag 
```
`cm` This flag allows to add the macros written specifically for the covergroup file. Complete command may look something like the following:
```objective-c
 riscv_isac --verbose info coverage -d                         -t /home/hammad/riscv-arch-test/riscof_work/torr_tests.S/torr_tests.log --parser-name c_sail -o coverage.rpt                          --sig-label begin_signature  end_signature  -e /home/hammad/riscv-arch-test/riscof_work/torr_tests.S/ref.elf -c /home/hammad/riscv-ctg/sample_cgfs/dataset.cgf -c /home/hammad/PMP_Coverpoints/checkcover_1.cgf -x32   -l virtual_memory_functions_test -h /home/hammad/header_file.yaml -cm self
```
Consider the following coverpoint:

```objective-c
  val_comb:
    ${shift_bit} == 1: 0
    ${final_value} == 0x012: 0
```

So, when user may run the above command, the coverpoints will be replaced by the values assigned in the headerfile written in `yaml` file.

allenjbaum commented on riscv_isac/coverage.py:
```diff
@@ -1006,23 +1005,52 @@ def get_mem_val(addr, bytes):
+                        return 1
+                    elif prop_name_lower == 'd' and (pte_per & 0x80 != 0):
+                        return 1
+                    elif prop_name_lower in {'rsw', 'RSW'} and (pte_per & 0x300 in {0, 1, 2, 3}):
```
This line doesn't look correct. IF you're checking to see if the result is in range {0,1,2,3} then you better have a mask in that range (0x3 - not 0x300) or you should make the range 0x000, 0x100, 0x200, 0x300).
And, regardless of that, if you're checking for 2 bits with 4 possible ranges, it will always match, regardless of the bit positions, because you're checking all possible 2 bit values.  Value 0 is whatyou're looking for, in which case you should check that it is !=0 instead of in {0,1,2,3}

allenjbaum commented on riscv_isac/coverage.py:
```diff
@@ -1006,23 +1005,52 @@ def get_mem_val(addr, bytes):
+                elif instr_vars['xlen'] == 32:
+                    pte_size = 4
                 if (pgtb_addr >> 12) == (pte_addr >> 12):
                     if ((pa >> 12) == (get_mem_val(pte_addr, pte_size) >> 10)):
```
Is the >>10 correct? Shouldn't it be 9 if it is SV39, 48, 57?

pawks commented on riscv_isac/InstructionObject.py:
```diff
@@ -316,6 +328,129 @@ def evaluate_instr_var(self, instr_var_name, *args):
+                    instr_vars[f'dptw{i}cont'] = None
+                    instr_vars[f'dptw{i}a'] = None
+            else:
+                raise Exception("Implemented Virtual Memory in the Test and the one given in the coverpoints is not same.")
```
I think this particular error message is misleading. If I understand correctly, this scenario should never occur as the SATP is tracked from the logs and should match the PT walk scheme being used.

MuhammadHammad001 replied:
Yes, thank you for the correction. I have pushed the corrected commit to this PR.

pawks commented on riscv_isac/coverage.py:
```diff
@@ -1065,6 +1146,16 @@ def write_fn_csr_comb_covpt(csr_reg):
+                                        match_imm = bool(re.search(pattern_imm, coverpoints))
+                                        match_rs1 = bool(re.search(pattern_rs1, coverpoints))
+                                        match_rs2 = bool(re.search(pattern_rs2, coverpoints))
+                                        if (instr_vars['rs2_val'] == None) and match_rs2:
```
I don't think the coverpoints for instructions without `rs2_val` and `imm_val` should be skipped. This should not be causing problems unless the variable is being indexed somewhere. Can you provide an example which causes this error to arise?

MuhammadHammad001 replied:
Please consider the example of following coverpoint, the instruction of interest will be picked by the **riscv-isac** using the following `mnemonics`.
```objective-c
  mnemonics:
    sw: 0
    csrrw: 0
    csrrs: 0
    csrrc: 0
    lw: 0
```
Also consider the following coverpoint written under the val_comb label. So, for the instructions like lw and sw which will obviously of interest because of the mnemonics will properly work for this coverpoint.
```objective-c
  val_comb:
    get_pte_prop("${U_bit}",get_addr("begin_signature"), rs1_val + imm_val, get_addr("rvtest_slvl1_pg_tbl")) == 1: 0

```
But, for the instructions which will have mnemonics like `csrrw`,`csrrs`,`csrrc` in them, will actually have no `imm_val` in the **instr_vars** dictionary and also the rs2_val will be of None Type. So, we can't perform any operations on them.

-------------------------------------------------------------------
One solution that I have tried to implement is to simply skip the instructions without imm_val and rs2_val.

The other solution can be to write the coverpoints in such a manner that they skip the instructions which do not contain the imm_val and rs2_val. But, for that I will have to change the following initialization of `imm_val` in the **Instruction_Object.py**.

https://github.com/riscv-software-src/riscv-isac/blob/db7504037ee611aa5b7e0b40e5207ae4871ef574/riscv_isac/InstructionObject.py#L159-L160

This will be changed such that the value of imm_val from the **c_sail.py** will always be added in the **instr_vars** even if its None so we can use the coverpoint properly with imm_val and rs2_val being of **None Type**

-----------------------------------------------------------------
So, for the example of coverpoint, we may write something like:
```objective-c
  val_comb:
    (get_pte_prop("${U_bit}",get_addr("begin_signature"), rs1_val + imm_val, get_addr("rvtest_slvl1_pg_tbl")) == 1) if imm_val is not None else False: 0

```

allenjbaum replied:
My understanding of this issue is that if they are not skipped, and are used as part of a computation (e.g. being indexed is one example), then they will fail. This can be fixed in one of two ways: explicitly skipped in the code that is defined for the instruction format, or modifying the .cgf files to effectively skip them. There are tradeoffs both ways, but the preference is to minimize the work, especially where there are future modifications to coverage (but not the formats, so explicitly skipping in the code was viewed as needing less work in the future.

pawks replied:
The grouping under `mnemonics` were only intended towards instructions which share coverpoints and hence the underlying architectural variables. If the underlying variables are not the same, they should not be sharing the same coverpoints and should be placed in different covergroups.

MuhammadHammad001 replied:
Yes, you are right, this was actually a short term solution, so I have removed this feature, so, in the coverpoints where we are targeting some specific instruction, we can do that by using the mnemonic flag(https://github.com/riscv-software-src/riscv-isac/pull/80#issue-1992987280) instead of changing the coverage.py for every instruction that does not have rs2_val and imm_val. So, the coverpoint look something like:
```
val_comb:
    mnemonic == ('sw' or 'lw') and (get_pte_prop("r",get_addr("begin_signature"), rs1_val + imm_val, get_addr("rvtest_slvl1_pg_tbl")) == 1): 0
```

pawks commented on riscv_isac/coverage.py:
```diff
@@ -957,17 +963,92 @@ def old_fn_csr_comb_covpt(csr_reg):
+                for csr_commit in instr.csr_commit:
+                    if(csr_commit[0] == "CSR") and (csr_commit[3]) and (csr_commit[2] == '->'):
+                        csr_read_vals[csr_commit[1]] = int(csr_commit[3],16)
+            def read_fn_csr_comb_covpt(csr_reg):
```
I think this function is redundant. The value for the csr being read is the same as the value in the arch state. Hence the coverpoints should not look different and follow the same format as `mcause == 1`. Something worth doing would be to check the value of the csr in the arch_state dictionary and verify it is correct.

MuhammadHammad001 replied:
This function is not redundant as the arch_state dictionary only takes the values that are written in the log. So, we faced a problem like for the `mcause` we did not know whether there was an exception or not because this register was never written in the log file with the exception. Therefore, to keep track of the values that were only read in the log which is differentiated using the `<-` or `->`, I had used this function.

So, earlier, the c_sail.py used to track only the registers that were written/overwritten in the log

https://github.com/riscv-software-src/riscv-isac/blob/db7504037ee611aa5b7e0b40e5207ae4871ef574/riscv_isac/plugins/c_sail.py#L20

But, now it takes care of both the registers that are read and also written

```python
    instr_pattern_c_sail_csr_reg_val = re.compile('(?P<CSR>CSR|clint::tick)\s(?P<reg>[a-z0-9]+)\s(.*?)\s(?P<val>[0-9xABCDEF]+)(?:\s\(input:\s(?P<input_val>[0-9xABCDEF]+)\))?')
```

allenjbaum replied:
In the absence of Sail explicitly generating a message for each update to architectural state (e.g. all the trap registers) we need something like this.
I have filed an issue with the Sail team to ensure that all implicit writes of CSRs appear in the Sail output (we will want some qualifications, e.g. minstret or mcycle should default to not logged, and maybe there are some categories we can selectively enable/disable

allenjbaum replied:
And to be explicit: until the Sail team ensures that all implicit writes of CSRs appear in the Sail output , we need to do this

pawks replied:
This is a stop-gap solution which ultimately introduces more issues because the meaning of the coverpoints get changed. This means that the coverpoints will have to be re-written to not include the `read_fn_csr_comb_covpt`. The primary fix would be to fix it in SAIL. A secondary fix would be to explictly overwrite the value being read from the log(indicated with `<-`) before evaluating and printing a warning that this feature would be deprecated in the future. Alternatively, just a warning to indicate that the value being read and the value in the logs are different. This check should stay on as a feature IMO.

MuhammadHammad001 replied:
Actually, there is a seperate track for the read_fn_csr_comb which is done by `csr_read_vals` dictionary, which is something like for a instruction lets say:
```
mem[X,0x80000A78] -> 0x23F3
mem[X,0x80000A7A] -> 0x3410
[202] [M]: 0x80000A78 (0x341023F3) csrrs t2, mepc, zero
CSR mepc -> 0x800003C0
x7 <- 0x800003C0
```
The value in the read_csr will be `csr_read_vals = {'mepc': 2147484608}` and this value is never tracked by either the arch_state nor the csr_regfile because it was never written by the user rather by the hart, so this can be accessed using the coverpoint:
```
csr_comb:
   read_csr("mepc") == 2147484608 : 0
```
This is just for an example. The main purpose for this functionality was to check the values for the mcause to check whether specific exception occured in the test or not as it will not be tracked by the arch_state or csr_regfile, but I have updated that as well(please refer to https://github.com/riscv-software-src/riscv-isac/pull/80#issuecomment-1910112010), so if you want me to remove this feature, please let me know.

MuhammadHammad001 replied:
This feature has been removed.

pawks commented on riscv_isac/coverage.py:
```diff
@@ -957,17 +963,92 @@ def old_fn_csr_comb_covpt(csr_reg):
+                if addr in mem_vals:
+                    return mem_vals.get(addr)
+                else:
+                    mem_val = utils.get_value_at_location(elf_path = elf, location = addr, bytes = bytes)
```
The value of the data should never be fetched from the elf as the elf would only contain the initial value. If a value is being used in the coverpoints, it should be obtained only from the logs. If the logs do not contain the value, they need to be fixed. Getting it from the elf might lead to false positives in coverage if the implementation is not correct.

MuhammadHammad001 replied:
If you please refer to **c_sail.py**, the function for extracting the values at every memory address from the log file for each instruction has been implemented. So, the flow of `mem_val` is such that it will first get the values of every address from the log files, and then, here in the **coverage.py** when the function `get_mem_val` is used, it first checks the `mem_val` and if this dictionary does not contain the value of that specific address mentioned in the coverpoint which will only happen if that address was initialized in the test(.S) but never used. Therefore, we will have the track of every address value whether updated or not by the log(s).

allenjbaum replied:
My interpretation is that this is just to ensure that statically initialized data gets initialized correctly. Do you see a problem with it?

pawks replied:
While, I see the value in ensuring that the statically initialized data is checking correctly, if the value is not loaded into the simulator properly i.e the (.S) files are not representative of the actual value loaded in SAIL we risk false positives. If the logs always contain the value being read(I believe this is already the case), I don't think there is any issue with not supporting this feature.

pawks commented on riscv_isac/main.py:
```diff
@@ -30,6 +31,18 @@ def cli(verbose):
+        type=click.Path(resolve_path=True,readable=True,exists=True),
+        help="YAML macro file to include"
+    )
+@click.option(
```
What are these CGF macros? What do they do?

MuhammadHammad001 replied:
Please refer to the [comment](https://github.com/riscv-software-src/riscv-isac/pull/80#issuecomment-1860541754). If you want me to change the naming format for these files/commands, please let me know. 
This allows you to add a custom macro file(.yaml based) that can be used to define your personal variables that are repeatedly used in the coverpoints. So, this is similar to `encoding.h` (riscv-arch-test).

pawks replied:
This looks reasonable. This and all the other features/variables added in this PR will need to be documented in the respective files/sections in `docs/source/` folder. For example all variable additions go in `docs/source/cgf.rst` with some examples similar to how the existing documentation describes the `op_comb`/`csr_comb` coverpoints.

MuhammadHammad001 replied:
@pawks , should I start working on the documentation, if you are okay with the implementation of other features?

pawks reviewed (dismissed):
The documentation should also be updated to indicate the added variables for the coverpoints.

Commit:
False exception removed for virtual memory support

Commit:
Trap Registers Variables added in instr_vars

MuhammadHammad001 commented:
# Trap Register Variables
I have added few variables that keep track of the current instruction exceptions/traps.
The variables are 
1. `mode_change` = From which mode we switched and to which mode. For instance, **S to M**
2. `exc_num` = Exception number
3. `tval` = tval
4. `call_type` = Type of exception. For instance **load-page-fault** or **e-call**

## Coverpoint example
User may write the coverpoints for these variables in the following pattern:
```objective-c
val_comb:
    call_type == "load-page-fault": 0
    exc_num == 15: 0
```
So, these are simple variables, user may change the above coverpoints depending upon the use-case.

allenjbaum commented:
I would have named it "Exc_type" rather than call type, (since it will be
confused with the Ecall instruction ABI parameter), but that's not
important.
Aren't call_type and exc_num redundant, that is, each call type is just an
ascii string that defines the exc_num?
It may not be the case (there may be different causes that have the same
exc_num, for example).
In any case - it doesn't hurt sof don't remove anything.

But, if you're exposing tval and exc_num (which I'm assuming is the cause
CSR) ... which tval, cause? Mtval/Mcause or Stval/Scause ? (and eventually
the hypervisor versions)
Not that if you're in Mmode, it code can still read&write stval.

We will need to add xtval2 & xtinst  eventually for hypervisor traps ,
though there is no hurry for that.




On Thu, Jan 18, 2024 at 11:09 AM HammadBashir ***@***.***>
wrote:

> Trap Register Variables
>
> I have added few variables that keep track of the current instruction
> exceptions/traps.
> The variables are
>
>    1. mode_change = From which mode we switched and to which mode. For
>    instance, *S to M*
>    2. exc_num = Exception number
>    3. tval = tval
>    4. call_type = Type of exception. For instance *load-page-fault* or
>    *e-call*
>
> Coverpoint example
>
> User may write the coverpoints for these variables in the following
> pattern:
>
> val_comb:
>     call_type == "load-page-fault": 0
>     exc_num == 15: 0
>
> So, these are simple variables, user may change the above coverpoints
> depending upon the use-case.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv-software-src/riscv-isac/pull/80#issuecomment-1899056737>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHPXVJTL3CKPRZK4EOWKTQ3YPFXP5AVCNFSM6AAAAAA7LAAW6OVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOJZGA2TMNZTG4>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Commit:
trap registers updated to maintain the arch-state

MuhammadHammad001 commented:
# Trap Register Variables
This is the updated comment as 
I have added few variables that keep track of the current instruction exceptions/traps.
The variables are 
1. `mode_change` = From which mode we switched and to which mode. For instance, **S to M**
2. `mcause/scause` = Exception number
3. `mtval/stval` = tval
4. `call_type` = Type of exception. For instance **load-page-fault** or **e-call**

## Coverpoint example
User may write the coverpoints for these variables in the following pattern:
```objective-c
val_comb:
    call_type == "load-page-fault": 0
    mcause == 15: 0
```

## Explanation
The code has been updated such that it will check from the mode_change whether the mode change was to supervisor or machine mode, then it will update the respective register i.e., in supervisor or machine mode register respectively. Please refer to the commit: https://github.com/riscv-software-src/riscv-isac/pull/80/commits/ac4f34251089db834c1aaf7274db8385b7f6f281

Now, the arch-state for these mcause/scause, mtval/stval is also maintained which means that their values will only be updated when they are changed in the log by the instruction or by a trap.

MuhammadHammad001 committed:
Merge branch 'dev' into read_csr_support

Signed-off-by: Muhammad Hammad Bashir <139617104+MuhammadHammad001@users.noreply.github.com>

MuhammadHammad001 commented:
@pawks , @allenjbaum  , @neelgala  
I would be really thankful if anyone please start the review process for this PR.

UmerShahidengr commented:
@allenjbaum,
@MuhammadHammad001 has resolved all conflicts in this PR, can we please merge it before any other conflict occurs?

UmerShahidengr commented:
@pawks,
your attention is also required here as this PR is stalled for a few weeks.

Commit:
Support for mnemonic flag and error resolved for inxflg

Commit:
Support for lr/sc added and also rd_val track added

Commit:
Remove the feature for rs2_val and imm_val skip for csrs in val_comb and update the unsgn_rs2 dict and evaluators for rd_val

Commit:
Feature for read_csr removed

MuhammadHammad001 commented:
@pawks All the required changes have been made, can you please review it again for any further changes

allenjbaum commented on riscv_isac/InstructionObject.py:
```diff
@@ -466,11 +476,28 @@ def trap_registers_update(self, instr_vars, trap_dict):
+            if trap_dict["mode_change"].split()[2] == "M":
+                instr_vars['mcause']      = trap_dict['exc_num']
+                instr_vars['mtval']       = trap_dict['tval']
+                instr_vars['scause']      = '0'
```
Stupid question: does this clear scause when the mode change is to Mmode? Because it should just leave it unchanged...

UmerShahidengr replied:
It will leave the scause unchanged, @MuhammadHammad001 can give more details but looking at the code, I think scause will remain unchanged

allenjbaum replied:
Well, I might believe the actual scause doesn't change. The question is whether ISAC's tracking of the scause value makes it think that it is now zero, so it messes up coverpoint matching (I think)

MuhammadHammad001 replied:
Yes, the values for mcause, scause, mtval and stval will be maintained until they are updated again by the exception/call.
P.S: I added the comment few days ago, but github made it into `pending`  which made it not visible to others.

Commit:
Update the trap registers state logic and add support for fetch access faults

Commit:
Initial Support for Translator added

MuhammadHammad001 commented:
# Introducing Translator Support
We've incorporated a script for translating CGF files to reduce redundancy at the user-end while writing the coverpoints. This script was developed in collaboration with Mr. @allenjbaum. To illustrate its functionality, let's examine a specific coverpoint:
```c-objective
PMP_TOR_priority_r:
  config:
    - check ISA:=regex(.*32.*); check ISA:=regex(.*I.*Zicsr.*);  def rvtest_mtrap_routine=True;
  mnemonics:
    "{csrrs, csrrw, lw, sw}" : 0
  csr_comb:
    ((pmpcfg3 >> 24) & {0x8C, 0x8A, 0x89}) == (0x8F & $1): 0                                  #Read, Write, execute Permission given to low priority region
    ((pmpcfg0 >> 24) & {0x8C, 0x8A, 0x89}) == (0x89 & $1): 0                                  #Read Permission given to high priority region
    pmpcfg{0, 3} != 0 and ((old("pmpcfg$1")) ^ (pmpcfg$1) != 0x00): 0                         #pmpcfg0(H.P0) and pmpcfg3(L.P) have been updated
    (old("pmpaddr{2, 3, 14, 15}")) ^ (pmpaddr$1) != 0x00: 0                                   #pmpaddr has been used and updated from the previous value i.e., 0x000
  val_comb:
    #Test for exceptions
    mode == {'M','S','U'} and (mcause != ${Load_access_fault}): 0                                             #No read fault
    mode == {'M','S','U'} and (mcause ==${Store_access_fault}): 0                                             #Write fault
    mode == 'M' and mode_change == {'M to M', 'U to M', 'S to M'} and (mcause == ${fetch_access_fault}): 0     #execute fault
    #check for the accesses
    ? (mnemonic == "lw" or mnemonic == "sw") and ((pmpcfg0 >> 24) & 0x9F == 0x89) and (rs1_val + imm_val >= (pmpaddr2 << 2)) and (rs1_val + imm_val < (pmpaddr3 << 2))
    : 0
```

The above will be processed by the translator_cgf.py and will result in the following coverpoint after the macros are resolved as well:

```c=objective
PMP_TOR_priority_r:
  config:
  - check ISA:=regex(.*32.*); check ISA:=regex(.*I.*Zicsr.*);  def rvtest_mtrap_routine=True;
  csr_comb:
    ((pmpcfg0 >> 24) & 0x89) == (0x89 & 0x89): 0
    ((pmpcfg0 >> 24) & 0x8A) == (0x89 & 0x8A): 0
    ((pmpcfg0 >> 24) & 0x8C) == (0x89 & 0x8C): 0
    ((pmpcfg3 >> 24) & 0x89) == (0x8F & 0x89): 0
    ((pmpcfg3 >> 24) & 0x8A) == (0x8F & 0x8A): 0
    ((pmpcfg3 >> 24) & 0x8C) == (0x8F & 0x8C): 0
    (old("pmpaddr14")) ^ (pmpaddr14) != 0x00: 0
    (old("pmpaddr15")) ^ (pmpaddr15) != 0x00: 0
    (old("pmpaddr2")) ^ (pmpaddr2) != 0x00: 0
    (old("pmpaddr3")) ^ (pmpaddr3) != 0x00: 0
    pmpcfg0 != 0 and ((old("pmpcfg0")) ^ (pmpcfg0) != 0x00): 0
    pmpcfg3 != 0 and ((old("pmpcfg3")) ^ (pmpcfg3) != 0x00): 0
  mnemonics:
    csrrs: 0
    csrrw: 0
    lw: 0
    sw: 0
  val_comb:
    ? (mnemonic == "lw" or mnemonic == "sw") and ((pmpcfg0 >> 24) & 0x9F == 0x89)
      and (rs1_val + imm_val >= (pmpaddr2 << 2)) and (rs1_val + imm_val < (pmpaddr3
      << 2))
    : 0
    mode == 'M' and (mcause != 0x05): 0
    mode == 'M' and (mcause == 0x07): 0
    mode == 'M' and mode_change == 'M to M' and (mcause == 0x01): 0
    mode == 'M' and mode_change == 'S to M' and (mcause == 0x01): 0
    mode == 'M' and mode_change == 'U to M' and (mcause == 0x01): 0
    mode == 'S' and (mcause != 0x05): 0
    mode == 'S' and (mcause == 0x07): 0
    mode == 'U' and (mcause != 0x05): 0
    mode == 'U' and (mcause == 0x07): 0
```

Commit:
Translator function added for merge coverage

Commit:
Documentation added for priv arch functions, translator, macros

UmerShahidengr committed:
Merge branch 'dev' into read_csr_support

Signed-off-by: Umer Shahid <umer.shahid@10xengineers.ai>

allenjbaum commented on riscv_isac/coverage.py:
```diff
@@ -977,6 +999,64 @@ def write_fn_csr_comb_covpt(csr_reg):
+                    pte_per = get_pte(pa, pte_addr, pgtb_addr) & 0x3FF
+                    prop_name_lower = prop_name.lower()
+                    if prop_name_lower == 'v' and (pte_per & 0x01 != 0):
+                        return 1
```
Don't we have constant definition we can use for all the different bit positions, e.g. #define pte_x_msk 0x08

allenjbaum replied:
In general, hex (or decimal) constants are bad, (zero is excluded from that broad statement.
You could also define the bit(variable, pos, length) function which extracts a field from a variable and right justifies it.

allenjbaum reviewed (approved):
This is a great addition to the functionality of creating coverpoints

allenjbaum reviewed (approved):
LGTM
